File: packages/data/src/scraping/boards/ycombinator.ts
/**
 * Y Combinator Jobs (Work at a Startup) — public Algolia search:
 *   https://hn.algolia.com/api/v1/search?tags=show_hn  -- (no, that's HN)
 * The reliable public path is the Algolia index that powers workatastartup search:
 *   POST https://0t3kibpncw-dsn.algolia.net/1/indexes/JobPosting/query
 *   with the WaaS public API key in headers.
 *
 * Keys rotate; treat failures gracefully. We fall back to an empty list.
 */
import type { JobPosting } from '@ajh/shared';
import { BaseScraper, type BoardSearchInput, type ScrapeContext } from '../base.js';
import { fetchJson, stripHtml } from '../http.js';

interface Hit {
  objectID: string;
  title?: string;
  description?: string;
  company_name?: string;
  location?: string;
  remote?: boolean;
  apply_url?: string;
  created_at_i?: number;
}
interface AlgoliaResp {
  hits: Hit[];
}

// Public Algolia creds for Work-at-a-Startup. If/when these rotate, this scraper
// degrades gracefully — main process never throws.
const ALGOLIA_APP_ID = '45BWZJ1SGC';
const ALGOLIA_API_KEY =
  'NDYyYmNmMDU5OWVmNzNlNzMwMWQ5MDE4ZWY3M2NlNDU0NjA5MTRmZTdiNDAxYjE3MTUyYmU5OWZlNjVmZmUyZHRhZ0ZpbHRlcnM9JTVCJTIyam9icyUyMiU1RA==';

export class YCombinatorScraper extends BaseScraper {
  readonly id = 'ycombinator';
  readonly displayName = 'Y Combinator';
  readonly mode = 'http' as const;

  async search(input: BoardSearchInput, ctx: ScrapeContext): Promise<JobPosting[]> {
    const q = input.query.trim();
    const url = `https://${ALGOLIA_APP_ID.toLowerCase()}-dsn.algolia.net/1/indexes/JobPosting/query`;
    const data = await fetchJson<AlgoliaResp>(url, {
      method: 'POST',
      headers: {
        'x-algolia-application-id': ALGOLIA_APP_ID,
        'x-algolia-api-key': ALGOLIA_API_KEY,
        'content-type': 'application/json',
      },
      body: JSON.stringify({ query: q, hitsPerPage: 50 }),
      ...(ctx.signal ? { signal: ctx.signal } : {}),
    });
    if (!data?.hits) return [];
    const now = Date.now();
    const out: JobPosting[] = [];
    for (const h of data.hits) {
      if (!h.title) continue;
      const posting: JobPosting = {
        id: this.makeId(h.objectID),
        source: this.id,
        externalId: h.objectID,
        url: h.apply_url ?? `https://www.workatastartup.com/jobs/${h.objectID}`,
        title: h.title,
        company: h.company_name ?? 'Unknown',
        location: h.location,
        remote: h.remote,
        description: stripHtml(h.description ?? ''),
        capturedAt: now,
        ...(h.created_at_i ? { postedAt: h.created_at_i * 1000 } : {}),
      };
      out.push(posting);
      await ctx.onItem?.(posting);
    }
    ctx.onProgress?.(1);
    return out;
  }
}

File: apps/tauri/src-tauri/src/scraping/boards/ycombinator.rs
/// Y Combinator Jobs (Work at a Startup) — public Algolia search.
use super::super::http::{fetch_json, strip_html};
use super::super::types::{BoardSearchInput, JobPosting, Scraper, ScraperMode, ScrapeContext};
use async_trait::async_trait;
use serde::Deserialize;

const ALGOLIA_APP_ID: &str = "45BWZJ1SGC";
const ALGOLIA_API_KEY: &str =
    "NDYyYmNmMDU5OWVmNzNlNzMwMWQ5MDE4ZWY3M2NlNDU0NjA5MTRmZTdiNDAxYjE3MTUyYmU5OWZlNjVmZmUyZHRhZ0ZpbHRlcnM9JTVCJTIyam9icyUyMiU1RA==";

#[derive(Debug, Deserialize)]
struct Hit {
    #[serde(rename = "objectID")]
    object_id: String,
    title: Option<String>,
    description: Option<String>,
    #[serde(rename = "company_name")]
    company_name: Option<String>,
    location: Option<String>,
    remote: Option<bool>,
    #[serde(rename = "apply_url")]
    apply_url: Option<String>,
    #[serde(rename = "created_at_i")]
    created_at_i: Option<i64>,
}

#[derive(Debug, Deserialize)]
struct AlgoliaResp {
    hits: Vec<Hit>,
}

pub struct YCombinatorScraper;

#[async_trait]
impl Scraper for YCombinatorScraper {
    fn id(&self) -> &'static str {
        "ycombinator"
    }

    fn display_name(&self) -> &'static str {
        "Y Combinator"
    }

    fn mode(&self) -> ScraperMode {
        ScraperMode::Http
    }

    async fn search(
        &self,
        input: BoardSearchInput,
        ctx: ScrapeContext,
    ) -> anyhow::Result<Vec<JobPosting>> {
        let q = input.query.trim();
        let url = format!(
            "https://{}-dsn.algolia.net/1/indexes/JobPosting/query",
            ALGOLIA_APP_ID.to_lowercase()
        );

        let data = fetch_json::<AlgoliaResp>(
            &url,
            super::super::http::FetchOptions {
                method: Some(reqwest::Method::POST),
                headers: Some(vec![
                    ("x-algolia-application-id".to_string(), ALGOLIA_APP_ID.to_string()),
                    ("x-algolia-api-key".to_string(), ALGOLIA_API_KEY.to_string()),
                    ("content-type".to_string(), "application/json".to_string()),
                ]),
                body: Some(serde_json::json!({ "query": q, "hitsPerPage": 50 }).to_string()),
                ..Default::default()
            },
            ctx.signal,
        )
        .await?;

        let hits = match data {
            Some(d) => d.hits,
            None => return Ok(vec![]),
        };

        let now = chrono::Utc::now().timestamp_millis();
        let mut out = vec![];

        for h in hits {
            if h.title.is_none() {
                continue;
            }

            let posting = JobPosting {
                id: format!("{}:{}", self.id(), h.object_id),
                external_id: Some(h.object_id.clone()),
                title: h.title.unwrap_or_default(),
                company: h.company_name.unwrap_or_else(|| "Unknown".to_string()),
                location: h.location,
                url: h.apply_url.unwrap_or_else(|| {
                    format!("https://www.workatastartup.com/jobs/{}", h.object_id)
                }),
                source: self.id().to_string(),
                description: h.description.map(|d| strip_html(&d)),
                requirements: None,
                posted_at: h.created_at_i.map(|t| t * 1000),
                captured_at: now,
                extra: {
                    let mut map = std::collections::HashMap::new();
                    if let Some(remote) = h.remote {
                        map.insert("remote".to_string(), serde_json::json!(remote));
                    }
                    map
                },
            };

            if let Some(ref on_item) = ctx.on_item {
                on_item(posting.clone());
            }

            out.push(posting);
        }

        if let Some(ref on_progress) = ctx.on_progress {
            on_progress(1.0);
        }

        Ok(out)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_ycombinator_scraper_id() {
        let scraper = YCombinatorScraper;
        assert_eq!(scraper.id(), "ycombinator");
    }

    #[test]
    fn test_ycombinator_scraper_display_name() {
        let scraper = YCombinatorScraper;
        assert_eq!(scraper.display_name(), "Y Combinator");
    }

    #[test]
    fn test_ycombinator_scraper_mode() {
        let scraper = YCombinatorScraper;
        assert_eq!(scraper.mode(), ScraperMode::Http);
    }
}
